webapp/blog/logic.py

```diff
@@ -16,7 +16,32 @@ def strip_excerpt(raw_html):
     return html.unescape(cleantext).replace("\n", "")
 
 
-def transform_article(article, featured_image=None, author=None):
+def replace_images_with_cloudinary(content):
+    """Prefixes images with cloudinary optimised URLs and adds srcset for
+    image scaling
+
+    :param content: The HTML string to convert
+
+    :returns: Update HTML string with converted images
+    """
+    cloudinary = (
+        "https://res.cloudinary.com/" "canonical/image/fetch/q_auto,f_auto,"
+    )
+    return re.sub(
+        r"img(.*)src=\"(.[^\"]*)\"",
+        r'img\1 src="{url}w_1120/\2"'
+        r'srcset="{url}w_750/\2 750w,'
+        r'{url}w_960/\2 960w, {url}w_1120/\2 1120w"'
+        r'sizes="(max-width: 375px) 560px,'
```

Comment:
Can someone explain this sizes attribute to me? 1) Why don't the suggested sizes here (`560px`, `880px`) match up to the sizes of images we're requesting from cloudinary? 2) Why are these widths so much smaller than the suggested image size? 3) How can we make any "sizes" recommendations when we don't know anything about the image we're replacing? Shouldn't we just use `sizes="100vw"`?

Reply:
Again I'll try your code and approve if it works!

With regards to the different sizes: It's a crude way to have sharper images even when optimised (pixels removed) by Cloudinary. I noticed some examples where the image that was sharp before became blurry - the most extreme of which was the infographic on https://snapcraft.io/blog/infographic-snaps-in-numbers.

I'm open to alternatives and I should probably study the documentation around `srcset` more closely

Reply:
Yeah I think there's clever stuff you can do with pixel density.

Reply:
:+1: will have a look, not sure there'll be something that fixes blurry images at 72dpi though :man_shrugging: